### PR TITLE
84 mep 3 add reinstall endpoint

### PR DESCRIPTION
**Pull request by kolsa**

### CODE REVIEW

File: cmd/metal-api/internal/datastore/switch.go [hunk 1]
```diff
@@ -71,7 +71,7 @@ func (rs *RethinkStore) SearchSwitchesConnectedToMachine(m *metal.Machine) ([]me
 		return nil, err
 	}
 
-	res := []metal.Switch{}
+	var res []metal.Switch
```
Comment (majst01) [inline cmd/metal-api/internal/datastore/switch.go]: unrelated

File: cmd/metal-api/internal/metal/ip_test.go [hunk 2]
```diff
@@ -51,7 +51,7 @@ func TestRemoveMachineId(t *testing.T) {
 		{
 			name:         "ip without machine tag",
 			ip:           IP{},
-			expectedTags: []string{},
+			expectedTags: nil,
```
Comment (majst01) [inline cmd/metal-api/internal/metal/ip_test.go]: unrelated

File: cmd/metal-api/internal/metal/ip_test.go [hunk 3]
```diff
@@ -72,7 +72,7 @@ func TestRemoveMachineId(t *testing.T) {
 			ip: IP{
 				Tags: []string{IpTag(tag.MachineID, "123")},
 			},
-			expectedTags: []string{},
+			expectedTags: nil,
```
Comment (majst01) [inline cmd/metal-api/internal/metal/ip_test.go]: unrelated

File: cmd/metal-api/internal/metal/ip_test.go [hunk 4]
```diff
@@ -110,7 +110,7 @@ func TestGetScope(t *testing.T) {
 			name: "project ip",
 			ip: IP{
 				ProjectID: "1",
-				Tags:      []string{},
+				Tags:      nil,
```
Comment (majst01) [inline cmd/metal-api/internal/metal/ip_test.go]: unrelated

File: cmd/metal-api/internal/metal/network.go [hunk 5]
```diff
@@ -50,7 +50,7 @@ func (p *Prefix) String() string {
 }
 
 func (p Prefixes) String() []string {
-	result := []string{}
+	var result []string
```
Comment (majst01) [inline cmd/metal-api/internal/metal/network.go]: unrelated

File: cmd/metal-api/internal/service/ip-service.go [hunk 6]
```diff
@@ -145,7 +145,7 @@ func (ir ipResource) listIPs(request *restful.Request, response *restful.Respons
 		return
 	}
 
-	result := []*v1.IPResponse{}
+	var result []*v1.IPResponse
```
Comment (majst01) [inline cmd/metal-api/internal/service/ip-service.go]: unrelated

File: cmd/metal-api/internal/service/ip-service.go [hunk 7]
```diff
@@ -169,7 +169,7 @@ func (ir ipResource) findIPs(request *restful.Request, response *restful.Respons
...
```
Comment (majst01) [inline cmd/metal-api/internal/service/ip-service.go]: unrelated

File: cmd/metal-api/internal/service/ip-service_test.go [hunk 8]
```diff
@@ -260,7 +260,7 @@ func TestUpdateIP(t *testing.T) {
 			wantedIPBase: &v1.IPBase{
 				ProjectID: testdata.IP1.ProjectID,
 				Type:      "static",
-				Tags:      []string{},
+				Tags:      nil,
```
Comment (majst01) [inline cmd/metal-api/internal/service/ip-service_test.go]: unrelated

File: cmd/metal-api/internal/service/machine-service.go [hunk 9]
```diff
@@ -92,7 +92,7 @@ func getPrivateNetwork(networks allocationNetworkMap) (*allocationNetwork, error
 
 // getMachineNetworks extracts the machines networks from an allocationNetworkMap
 func getMachineNetworks(networks allocationNetworkMap) []*metal.MachineNetwork {
-	machineNetworks := []*metal.MachineNetwork{}
+	var machineNetworks []*metal.MachineNetwork
```
Comment (majst01) [inline cmd/metal-api/internal/service/machine-service.go]: unrelated

File: cmd/metal-api/internal/service/machine-service.go [hunk 10]
```diff
@@ -1311,7 +1336,7 @@ func makeMachineNetwork(ds *datastore.RethinkStore, ipamer ipam.IPAMer, allocati
 		n.ips = append(n.ips, *ip)
 	}
 
-	ipAddresses := []string{}
+	var ipAddresses []string
```
Comment (majst01) [inline cmd/metal-api/internal/service/machine-service.go]: unrelated

File: cmd/metal-api/internal/service/machine-service.go [hunk 11]
```diff
@@ -1382,10 +1407,10 @@ func makeMachineTags(m *metal.Machine, networks allocationNetworkMap, userTags [
 
 	// as user labels are given as an array, we need to figure out if label-like tags were provided.
 	// otherwise the user could provide confusing information like:
-	// - machine.metal-stack.io/chassis=123
-	// - machine.metal-stack.io/chassis=789
+	// - machine.metal-pod.io/chassis=123
```
Comment (majst01) [inline cmd/metal-api/internal/service/machine-service.go]: was already fixed in master to metal-stack

File: cmd/metal-api/internal/service/machine-service.go [hunk 12]
```diff
@@ -1382,10 +1407,10 @@ func makeMachineTags(m *metal.Machine, networks allocationNetworkMap, userTags [
...
+	// - machine.metal-pod.io/chassis=789
 	userLabels := make(map[string]string)
-	actualUserTags := []string{}
+	var actualUserTags []string
```
Comment (majst01) [inline cmd/metal-api/internal/service/machine-service.go]: unrelated

File: cmd/metal-api/internal/service/machine-service.go [hunk 13]
```diff
@@ -1435,7 +1460,7 @@ func uniqueTags(tags []string) []string {
 	for _, t := range tags {
 		tagSet[t] = true
 	}
-	uniqueTags := []string{}
+	var uniqueTags []string
```
Comment (majst01) [inline cmd/metal-api/internal/service/machine-service.go]: unrelated

File: cmd/metal-api/internal/service/machine-service.go [hunk 14]
```diff
@@ -1895,7 +1982,7 @@ func (r machineResource) machineCmd(op string, cmd metal.MachineCommand, request
 }
 
 func publishMachineCmd(logger *zap.SugaredLogger, m *metal.Machine, publisher bus.Publisher, cmd metal.MachineCommand, params ...string) error {
-	pp := []string{}
+	var pp []string
```
Comment (majst01) [inline cmd/metal-api/internal/service/machine-service.go]: unrelated

File: cmd/metal-api/internal/service/machine-service.go [hunk 15]
```diff
@@ -1948,7 +2035,7 @@ func makeMachineResponse(m *metal.Machine, ds *datastore.RethinkStore, logger *z
 func makeMachineResponseList(ms metal.Machines, ds *datastore.RethinkStore, logger *zap.SugaredLogger) []*v1.MachineResponse {
 	sMap, pMap, iMap, ecMap := getMachineReferencedEntityMaps(ds, logger)
 
-	result := []*v1.MachineResponse{}
+	var result []*v1.MachineResponse
```
Comment (majst01) [inline cmd/metal-api/internal/service/machine-service.go]: unrelated

File: cmd/metal-api/internal/service/machine-service.go [hunk 16]
```diff
@@ -1983,7 +2070,7 @@ func makeMachineIPMIResponse(m *metal.Machine, ds *datastore.RethinkStore, logge
 func makeMachineIPMIResponseList(ms metal.Machines, ds *datastore.RethinkStore, logger *zap.SugaredLogger) []*v1.MachineIPMIResponse {
 	sMap, pMap, iMap, ecMap := getMachineReferencedEntityMaps(ds, logger)
 
-	result := []*v1.MachineIPMIResponse{}
+	var result []*v1.MachineIPMIResponse
```
Comment (majst01) [inline cmd/metal-api/internal/service/machine-service.go]: unrelated

File: cmd/metal-api/internal/service/network-service.go [hunk 17]
```diff
@@ -184,7 +184,7 @@ func (r networkResource) findNetworks(request *restful.Request, response *restfu
 		return
 	}
 
-	result := []*v1.NetworkResponse{}
+	var result []*v1.NetworkResponse
```
Comment (majst01) [inline cmd/metal-api/internal/service/network-service.go]: unrelated

File: cmd/metal-api/internal/service/partition-service.go [hunk 18]
```diff
@@ -135,7 +135,7 @@ func (r partitionResource) listPartitions(request *restful.Request, response *re
 		return
 	}
 
-	result := []*v1.PartitionResponse{}
+	var result []*v1.PartitionResponse
```
Comment (majst01) [inline cmd/metal-api/internal/service/partition-service.go]: unrelated

File: cmd/metal-api/internal/service/partition-service.go [hunk 19]
```diff
@@ -320,9 +320,8 @@ func (r partitionResource) calcPartitionCapacity() ([]v1.PartitionCapacity, erro
 	}
 	machines := makeMachineResponseList(ms, r.ds, zapup.MustRootLogger().Sugar())
 
-	partitionCapacities := []v1.PartitionCapacity{}
+	var partitionCapacities []v1.PartitionCapacity
```
Comment (majst01) [inline cmd/metal-api/internal/service/partition-service.go]: unrelated

File: cmd/metal-api/internal/service/partition-service.go [hunk 20]
```diff
@@ -361,10 +360,12 @@ func (r partitionResource) calcPartitionCapacity() ([]v1.PartitionCapacity, erro
 
 			cap.Total++
 		}
-
-		sc := []v1.ServerCapacity{}
-		for i := range capacities {
-			sc = append(sc, *capacities[i])
+		var sc []v1.ServerCapacity
```
Comment (majst01) [inline cmd/metal-api/internal/service/partition-service.go]: unrelated

File: cmd/metal-api/internal/service/size-service.go [hunk 21]
```diff
@@ -120,7 +120,7 @@ func (r sizeResource) listSizes(request *restful.Request, response *restful.Resp
 		return
 	}
 
-	result := []*v1.SizeResponse{}
+	var result []*v1.SizeResponse
```
Comment (majst01) [inline cmd/metal-api/internal/service/size-service.go]: unrelated

File: cmd/metal-api/internal/service/switch-service.go [hunk 22]
```diff
@@ -336,8 +336,7 @@ func makeSwitchResponse(s *metal.Switch, ds *datastore.RethinkStore, logger *zap
 }
 
 func makeBGPFilterFirewall(m metal.Machine) v1.BGPFilter {
-	vnis := []string{}
-	cidrs := []string{}
+	var vnis, cidrs []string
```
Comment (majst01) [inline cmd/metal-api/internal/service/switch-service.go]: unrelated

File: cmd/metal-api/internal/service/switch-service.go [hunk 23]
```diff
@@ -352,8 +351,7 @@ func makeBGPFilterFirewall(m metal.Machine) v1.BGPFilter {
 }
 
 func makeBGPFilterMachine(m metal.Machine, ips metal.IPsMap) v1.BGPFilter {
-	vnis := []string{}
-	cidrs := []string{}
+	var vnis, cidrs []string
```
Comment (majst01) [inline cmd/metal-api/internal/service/switch-service.go]: unrelated

File: cmd/metal-api/internal/service/switch-service.go [hunk 24]
```diff
@@ -432,7 +430,7 @@ func makeSwitchNics(s *metal.Switch, ips metal.IPsMap, iMap metal.ImageMap, mach
 }
 
 func makeSwitchCons(s *metal.Switch) []v1.SwitchConnection {
-	cons := []v1.SwitchConnection{}
+	var cons []v1.SwitchConnection
```
Comment (majst01) [inline cmd/metal-api/internal/service/switch-service.go]: unrelated

File: cmd/metal-api/internal/service/switch-service.go [hunk 25]
```diff
@@ -482,7 +480,7 @@ func findSwitchReferencedEntites(s *metal.Switch, ds *datastore.RethinkStore, lo
 
 func makeSwitchResponseList(ss []metal.Switch, ds *datastore.RethinkStore, logger *zap.SugaredLogger) []*v1.SwitchResponse {
 	pMap, ips, iMap := getSwitchReferencedEntityMaps(ds, logger)
-	result := []*v1.SwitchResponse{}
+	var result []*v1.SwitchResponse
```
Comment (majst01) [inline cmd/metal-api/internal/service/switch-service.go]: unrelated

File: cmd/metal-api/internal/service/switch-service_test.go [hunk 26]
```diff
@@ -231,7 +231,7 @@ func TestMakeBGPFilterFirewall(t *testing.T) {
 					Allocation: &metal.MachineAllocation{
 						MachineNetworks: []*metal.MachineNetwork{
 							&metal.MachineNetwork{
-								IPs: []string{},
+								IPs: nil,
```
Comment (majst01) [inline cmd/metal-api/internal/service/switch-service_test.go]: unrelated

File: cmd/metal-api/internal/service/switch-service_test.go [hunk 27]
```diff
@@ -266,7 +266,7 @@ func TestMakeBGPFilterFirewall(t *testing.T) {
 			},
 			want: v1.BGPFilter{
 				VNIs:  []string{"104010"},
-				CIDRs: []string{},
+				CIDRs: nil,
```
Comment (majst01) [inline cmd/metal-api/internal/service/switch-service_test.go]: unrelated

File: cmd/metal-api/internal/service/switch-service_test.go [hunk 28]
```diff
@@ -279,8 +279,8 @@ func TestMakeBGPFilterFirewall(t *testing.T) {
 				},
 			},
 			want: v1.BGPFilter{
-				VNIs:  []string{},
-				CIDRs: []string{},
+				VNIs:  nil,
```
Comment (majst01) [inline cmd/metal-api/internal/service/switch-service_test.go]: unrelated

File: cmd/metal-api/internal/service/switch-service_test.go [hunk 29]
```diff
@@ -344,7 +344,7 @@ func TestMakeBGPFilterMachine(t *testing.T) {
 					},
 				},
 			},
-			want: v1.NewBGPFilter([]string{}, []string{"10.1.0.0/22", "10.2.0.0/22", "100.127.1.1/32", "212.89.42.1/32", "212.89.42.2/32"}),
+			want: v1.NewBGPFilter(nil, []string{"10.1.0.0/22", "10.2.0.0/22", "100.127.1.1/32", "212.89.42.1/32", "212.89.42.2/32"}),
```
Comment (majst01) [inline cmd/metal-api/internal/service/switch-service_test.go]: unrelated

File: cmd/metal-api/internal/service/switch-service_test.go [hunk 30]
```diff
@@ -366,7 +366,7 @@ func TestMakeBGPFilterMachine(t *testing.T) {
 					},
 				},
 			},
-			want: v1.NewBGPFilter([]string{}, []string{"212.89.42.1/32"}),
+			want: v1.NewBGPFilter(nil, []string{"212.89.42.1/32"}),
```
Comment (majst01) [inline cmd/metal-api/internal/service/switch-service_test.go]: unrelated

File: cmd/metal-api/internal/service/switch-service_test.go [hunk 31]
```diff
@@ -462,14 +462,14 @@ func TestMakeSwitchNics(t *testing.T) {
 					Vrf:  "vrf1",
 					BGPFilter: &v1.BGPFilter{
 						CIDRs: []string{"212.89.1.1/32"},
-						VNIs:  []string{},
+						VNIs:  nil,
```
Comment (majst01) [inline cmd/metal-api/internal/service/switch-service_test.go]: unrelated

File: cmd/metal-api/internal/service/switch-service_test.go [hunk 32]
```diff
@@ -462,14 +462,14 @@ func TestMakeSwitchNics(t *testing.T) {
...
 					},
 				},
 				v1.SwitchNic{
 					Name: "swp2",
 					Vrf:  "default",
 					BGPFilter: &v1.BGPFilter{
-						CIDRs: []string{},
+						CIDRs: nil,
```
Comment (majst01) [inline cmd/metal-api/internal/service/switch-service_test.go]: unrelated

File: cmd/metal-api/internal/service/v1/image.go [hunk 33]
```diff
@@ -30,7 +30,7 @@ func NewImageResponse(img *metal.Image) *ImageResponse {
 	if img == nil {
 		return nil
 	}
-	features := []string{}
+	var features []string
```
Comment (majst01) [inline cmd/metal-api/internal/service/v1/image.go]: unrelated

File: cmd/metal-api/internal/service/v1/machine.go [hunk 34]
```diff
@@ -205,8 +240,13 @@ type MachineIpmiReportResponse struct {
...
 }
 
+type MachineReinstallRequest struct {
+	Common
+	ImageID string `json:"imageid" description:"the image id to be installed"`
+}
+
 func NewMetalMachineHardware(r *MachineHardwareExtended) metal.MachineHardware {
-	nics := metal.Nics{}
+	var nics metal.Nics
```
Comment (majst01) [inline cmd/metal-api/internal/service/v1/machine.go]: unrelated

File: cmd/metal-api/internal/service/v1/machine.go [hunk 35]
```diff
@@ -322,7 +378,7 @@ func NewMachineIPMIResponse(m *metal.Machine, s *metal.Size, p *metal.Partition,
 
 func NewMachineResponse(m *metal.Machine, s *metal.Size, p *metal.Partition, i *metal.Image, ec *metal.ProvisioningEventContainer) *MachineResponse {
 	var hardware MachineHardware
-	nics := MachineNics{}
+	var nics MachineNics
```
Comment (majst01) [inline cmd/metal-api/internal/service/v1/machine.go]: unrelated

File: cmd/metal-api/internal/service/v1/machine.go [hunk 36]
```diff
@@ -350,20 +406,19 @@ func NewMachineResponse(m *metal.Machine, s *metal.Size, p *metal.Partition, i *
...
-				Underlay:            m.Allocation.MachineNetworks[i].Underlay,
-				DestinationPrefixes: m.Allocation.MachineNetworks[i].DestinationPrefixes,
-				Prefixes:            m.Allocation.MachineNetworks[i].Prefixes,
+				Vrf:                 nw.Vrf,
+				ASN:                 nw.ASN,
+				Private:             nw.Private,
+				Nat:                 nw.Nat,
+				Underlay:            nw.Underlay,
+				DestinationPrefixes: nw.DestinationPrefixes,
+				Prefixes:            nw.Prefixes,
```
Comment (majst01) [inline cmd/metal-api/internal/service/v1/machine.go]: unrelated

File: cmd/metal-api/internal/service/v1/machine.go [hunk 37]
```diff
@@ -386,9 +441,21 @@ func NewMachineResponse(m *metal.Machine, s *metal.Size, p *metal.Partition, i *
...
+				OSPartition:  m.Allocation.OSPartition,
+				Initrd:       m.Allocation.Initrd,
+				Cmdline:      m.Allocation.Cmdline,
+				Kernel:       m.Allocation.Kernel,
+				BootloaderID: m.Allocation.BootloaderID,
+			}
+		}
 	}
 
-	tags := []string{}
```
Comment (majst01) [inline cmd/metal-api/internal/service/v1/machine.go]: unrelated

File: cmd/metal-api/internal/service/v1/machine.go [hunk 38]
```diff
@@ -438,7 +505,7 @@ func NewMachineResponse(m *metal.Machine, s *metal.Size, p *metal.Partition, i *
 }
 
 func NewMachineRecentProvisioningEvents(ec *metal.ProvisioningEventContainer) *MachineRecentProvisioningEvents {
-	es := []MachineProvisioningEvent{}
+	var es []MachineProvisioningEvent
```
Comment (majst01) [inline cmd/metal-api/internal/service/v1/machine.go]: unrelated

File: cmd/metal-api/internal/service/v1/size.go [hunk 39]
```diff
@@ -40,7 +40,7 @@ type SizeMatchingLog struct {
 }
 
 func NewSizeMatchingLog(m *metal.SizeMatchingLog) *SizeMatchingLog {
-	constraints := []SizeConstraintMatchingLog{}
+	var constraints []SizeConstraintMatchingLog
```
Comment (majst01) [inline cmd/metal-api/internal/service/v1/size.go]: unrelated

File: cmd/metal-api/internal/tags/tags.go [hunk 40]
```diff
@@ -44,7 +44,7 @@ func (t *Tags) Add(tag string) {
 
 // Remove removes a tag
 func (t *Tags) Remove(tag string) bool {
-	tags := []string{}
+	var tags []string
```
Comment (majst01) [inline cmd/metal-api/internal/tags/tags.go]: unrelated

File: cmd/metal-api/internal/tags/tags.go [hunk 41]
```diff
@@ -61,7 +61,7 @@ func (t *Tags) Remove(tag string) bool {
 
 // Values collects all the values that are contained with the given prefix.
 func (t *Tags) Values(prefix string) []string {
-	values := []string{}
+	var values []string
```
Comment (majst01) [inline cmd/metal-api/internal/tags/tags.go]: unrelated

File: cmd/metal-api/internal/tags/tags.go [hunk 42]
```diff
@@ -76,7 +76,7 @@ func (t *Tags) Unique() []string {
 	for _, t := range t.tags {
 		tagSet[t] = true
 	}
-	uniqueTags := []string{}
+	var uniqueTags []string
```
Comment (majst01) [inline cmd/metal-api/internal/tags/tags.go]: unrelated

File: cmd/metal-api/internal/tags/tags_test.go [hunk 43]
```diff
@@ -27,7 +27,7 @@ func TestHas(t *testing.T) {
 	}{
 		{
 			name:   "empty",
-			tags:   []string{},
+			tags:   nil,
```
Comment (majst01) [inline cmd/metal-api/internal/tags/tags_test.go]: unrelated

File: cmd/metal-api/internal/tags/tags_test.go [hunk 44]
```diff
@@ -63,7 +63,7 @@ func TestHasPrefix(t *testing.T) {
 	}{
 		{
 			name:   "empty tags",
-			tags:   []string{},
+			tags:   nil,
```
Comment (majst01) [inline cmd/metal-api/internal/tags/tags_test.go]: unrelated

File: cmd/metal-api/internal/tags/tags_test.go [hunk 45]
```diff
@@ -136,8 +136,8 @@ func TestUnique(t *testing.T) {
 	}{
 		{
 			name:   "empty",
-			tags:   []string{},
-			wanted: []string{},
+			tags:   nil,
```
Comment (majst01) [inline cmd/metal-api/internal/tags/tags_test.go]: unrelated

File: cmd/metal-api/internal/service/machine-service.go [hunk 46]
```diff
@@ -1462,7 +1487,15 @@ func (r machineResource) finalizeAllocation(request *restful.Request, response *
 	}
 
 	old := *m
+
 	m.Allocation.ConsolePassword = requestPayload.ConsolePassword
+	m.Allocation.PrimaryDisk = requestPayload.PrimaryDisk
```
Comment (majst01) [inline cmd/metal-api/internal/service/machine-service.go]: What happens for already allocated machines which have been allocated before this feature was implemented ?

**Comment by majst01** [comment]
Can you please resolve the merge conflicts, other than that LGTM. 
Also a retest of the current state of api/core/hammer in the metal-lab seems required before going to real.

**Comment by kolsa** [comment]
Merged and successfully retested in mini-lab

**Comment by majst01** [comment]
OK, @Gerrit91 should also have a look, can you meanwhile squash the commits to one

File: cmd/metal-api/internal/metal/provisioning.go [hunk 47]
```diff
@@ -41,6 +42,7 @@ var (
 		ProvisioningEventInstalling:       true,
 		ProvisioningEventBootingNewKernel: true,
 		ProvisioningEventPhonedHome:       true,
+		ProvisioningEventReinstallAborted: true,
```
Comment (Gerrit91) [inline cmd/metal-api/internal/metal/provisioning.go]: This will actually indicate for the state machine that this is the "final state" of machine provisioning. The reinstall should not be part of the regular provisioning sequence. Instead, it should go into the `expectedSuccessorEventMap` where you can define from which state it is allowed to transition into reinstall and to which state it is allowed to proceed after the reinstall.

It would also be nice to have a test for that. You can just use example sequences of provisioning events and check whether the state machine behaves as expected or not.

File: cmd/metal-api/internal/service/machine-service.go [hunk 48]
```diff
@@ -269,6 +269,17 @@ func (r machineResource) webService() *restful.WebService {
...
 
+	ws.Route(ws.POST("/{id}/reinstall").
+		To(editor(r.reinstallMachine)).
+		Operation("reinstallMachine").
+		Doc("reinstall this machine").
+		Param(ws.PathParameter("id", "identifier of the machine").DataType("string")).
+		Metadata(restfulspec.KeyOpenAPITags, tags).
+		Reads(v1.MachineReinstallRequest{}).
+		Returns(http.StatusOK, "OK", v1.MachineResponse{}).
+		Returns(http.StatusGatewayTimeout, "Timeout", httperrors.HTTPErrorResponse{}).
```
Comment (Gerrit91) [inline cmd/metal-api/internal/service/machine-service.go]: This endpoint does not return a status gateway timeout. I guess, you copied that from the wait endpoint. If this is the case, you can remove this line.

File: cmd/metal-api/internal/service/machine-service.go [hunk 49]
```diff
@@ -1493,31 +1524,38 @@ func (r machineResource) finalizeAllocation(request *restful.Request, response *
...
-	err = response.WriteHeaderAndEntity(http.StatusOK, makeMachineResponse(m, r.ds, logger))
+	err = response.WriteHeaderAndEntity(http.StatusOK, makeMachineResponse(m, r.ds, utils.Logger(request).Sugar()))
 	if err != nil {
 		logger.Error("Failed to send response", zap.Error(err))
-		return
 	}
 }
 
-func freeMachine(ds *datastore.RethinkStore, publisher bus.Publisher, ipamer ipam.IPAMer, m *metal.Machine, logger *zap.SugaredLogger) error {
+func reinstallOrDeleteMachine(ds *datastore.RethinkStore, publisher bus.Publisher, ipamer ipam.IPAMer, m *metal.Machine, imageID *string, logger *zap.SugaredLogger) error {
```
Comment (Gerrit91) [inline cmd/metal-api/internal/service/machine-service.go]: To me it looks like the original `freeMachine` function and `reinstallMachine` do not share enough code to justify combining them with a more complex function `reinsintallOrDeleteMachine`, adding an additional parameter to the function (completely unnecessary for `freeMachine`) and more conditional branches in the function body. To my taste it would make it a little easier to understand the procedure if there was a separate method `reinstallMachine` and the `freeMachine` method stays untouched. I would also say it is more future proof to separate this functionality right from the beginning. What do you think?

File: cmd/metal-api/internal/service/machine-service.go [hunk 50]
```diff
@@ -1532,27 +1570,50 @@ func freeMachine(ds *datastore.RethinkStore, publisher bus.Publisher, ipamer ipa
...
 	}
-	logger.Infow("freed machine", "machineID", m.ID)
 
-	return nil
+	logger.Infow(action, "machineID", m.ID)
+
+	if imageID != nil {
+		err = publishMachineCmd(logger, m, publisher, metal.MachineReinstall)
+		if err != nil {
+			logger.Errorw("unable to publish ’Reinstall' command", "machineID", m.ID, "error", err)
```
Comment (Gerrit91) [inline cmd/metal-api/internal/service/machine-service.go]: ’Reinstall' -> mixed apostrophe and backtick

File: cmd/metal-api/internal/service/machine-service.go [hunk 51]
```diff
@@ -1690,6 +1751,8 @@ func (r machineResource) provisioningEventForMachine(machineID string, e v1.Mach
 	} else if event.Event == metal.ProvisioningEventPhonedHome && len(ec.Events) > 0 && ec.Events[0].Event == metal.ProvisioningEventPhonedHome {
 		zapup.MustRootLogger().Sugar().Debugw("swallowing repeated phone home event", "id", ec.ID)
 		ec.Liveliness = metal.MachineLivelinessAlive
+	} else if event.Event == metal.ProvisioningEventReinstallAborted {
```
Comment (Gerrit91) [inline cmd/metal-api/internal/service/machine-service.go]: This way you swallow the event, which is "ok". But maybe it would be nice to track it in the state machine because then a user could see what went wrong in machine describe. And it would also be added to the machine issues, which could be beneficial for operations.

File: cmd/metal-api/internal/service/machine-service.go [hunk 52]
```diff
@@ -1701,6 +1764,30 @@ func (r machineResource) provisioningEventForMachine(machineID string, e v1.Mach
...
+		m.Allocation.Reinstall = false
+		err = r.ds.UpdateMachine(&old, m)
+		if err != nil {
+			log.Errorw("unable to find machine", "machineID", machineID, "error", err)
+		}
+	}
+
+	err = publishMachineCmd(log, m, r, metal.MachineAbortReinstall)
+	if err != nil {
+		log.Errorw("unable to publish ’Abort Reinstall' command", "machineID", machineID, "error", err)
```
Comment (Gerrit91) [inline cmd/metal-api/internal/service/machine-service.go]: Apostrophe and backtick mixed

File: cmd/metal-api/internal/service/partition-service.go [hunk 53]
```diff
@@ -359,9 +358,11 @@ func (r partitionResource) calcPartitionCapacity() ([]v1.PartitionCapacity, erro
 
 			cap.Total++
 		}
-
 		sc := []v1.ServerCapacity{}
 		for i := range capacities {
+			if capacities[i] == nil {
```
Comment (Gerrit91) [inline cmd/metal-api/internal/service/partition-service.go]: Oh, interesting! This seems to be fixing a bug but also seems unrelated to this PR. Can we file this in another issue and a separate PR? :)

File: cmd/metal-api/internal/service/machine-service.go [hunk 54]
```diff
@@ -1690,6 +1751,8 @@ func (r machineResource) provisioningEventForMachine(machineID string, e v1.Mach
 	} else if event.Event == metal.ProvisioningEventPhonedHome && len(ec.Events) > 0 && ec.Events[0].Event == metal.ProvisioningEventPhonedHome {
 		zapup.MustRootLogger().Sugar().Debugw("swallowing repeated phone home event", "id", ec.ID)
 		ec.Liveliness = metal.MachineLivelinessAlive
+	} else if event.Event == metal.ProvisioningEventReinstallAborted {
+		r.machineAbortReinstall(machineID)
```
Comment (Gerrit91) [inline cmd/metal-api/internal/service/machine-service.go]: This is kind of a minor nit: I am not sure when "aborting the reinstall" happens, but I would let it call the metal-core directly on a dedicated endpoint and not indirectly via a machine event. This endpoint was built in a way that it does only deal with the event table. Now it's introducing more complex things.

File: cmd/metal-api/internal/service/v1/machine.go [hunk 55]
```diff
@@ -380,6 +435,20 @@ func NewMachineResponse(m *metal.Machine, s *metal.Size, p *metal.Partition, i *
 			MachineNetworks: networks,
 			Succeeded:       m.Allocation.Succeeded,
 		}
+
+		if m.Allocation.Reinstall {
```
Comment (Gerrit91) [inline cmd/metal-api/internal/service/v1/machine.go]: So... only during reinstallation you have the MachineSetup fields available and when it succeeded this is gone? Aren't these fields true before and after a reinstall?

Comment (Gerrit91) [reply cmd/metal-api/internal/service/v1/machine.go]: Just make sure that this is only a mapping function. It should be dumb.

File: spec/metal-api.json [hunk 56]
```diff
@@ -1,5494 +1,5729 @@
...
-                },
-                "value": {
-                    "description": "the state of this machine. empty means available for all",
-                    "type": "string"
-                }
-            },
-            "required": [
-                "description",
-                "value"
-
```
Comment (Gerrit91) [inline spec/metal-api.json]: Huh? Why is this file completely different now? Where is the difference between the `jq` magic and the Python magic?

File: .github/workflows/branches.yaml [hunk 57]
```diff
@@ -0,0 +1,26 @@
+name: Build from branches except master
```
[the remark this answers is not in the document]
Comment (majst01) [reply .github/workflows/branches.yaml]: Can we remove this as we merged #43

[the remark this answers is not in the document]
Comment (majst01) [reply .github/workflows/branches.yaml]: If this succeeded we will have a metal-api image with a tag from this PR

File: cmd/metal-api/internal/service/machine-service.go [hunk 58]
```diff
@@ -1508,51 +1528,118 @@ func (r machineResource) freeMachine(request *restful.Request, response *restful
...
 		err = releaseMachineNetworks(ds, ipamer, m)
 		if err != nil {
 			// TODO: Trigger network garbage collection
 			// TODO: Check if all IPs in rethinkdb are in the IPAM and vice versa, cleanup if this is not the case
 			// TODO: Check if there are network prefixes in the IPAM that are not in any of our networks
-			logger.Errorw("an error during releasing machine networks occurred, scheduled network garbage collection", "error", err)
+			logger.Errorf("an error during releasing machine networks occurred, scheduled network garbage collection", "error", err)
+		}
+
+		old := *m
```
Comment (Gerrit91) [inline cmd/metal-api/internal/service/machine-service.go]: Oops, you put that into the if condition, but this needs to be outside the condition... just as before.

File: cmd/metal-api/internal/service/machine-service.go [hunk 59]
```diff
@@ -1555,6 +1578,129 @@ func freeMachine(ds *datastore.RethinkStore, publisher bus.Publisher, ipamer ipa
...
+	if checkError(request, response, utils.CurrentFuncName(), err) {
+		return
+	}
+
+	logger := utils.Logger(request).Sugar()
+
+	var resp *v1.MachineResponse
+	if m.Allocation != nil {
+		old := *m
+
```
Comment (majst01) [inline cmd/metal-api/internal/service/machine-service.go]: we should also check here if m.Allocation.MachineSetup is nil, if yes reinstallation can not be done right?

**Comment by kolsa** [comment]
We could do so, but then we would prevent reinstalling machines that are not yet affected by the new changes.  
But why not allowing this too? If no error occurs during reinstallation the machine will get the missing machine setup.  
In contrary, if the reinstallation fails for any reason, then yes, the required bootinfo cannot be constructed and the machine will not be able to reboot again. It rather falls into a reinstallation loop until the reinstall will succeed, eventually

**Comment by majst01** [comment]
OK, understood, the extra check can be added later anyway.

File: cmd/metal-api/internal/service/machine-service.go [hunk 60]
```diff
@@ -1508,44 +1540,35 @@ func (r machineResource) freeMachine(request *restful.Request, response *restful
...
+	// if the machine is allocated, we free it in our database
 	if m.Allocation != nil {
-		// if the machine is allocated, we free it in our database
 		err = releaseMachineNetworks(ds, ipamer, m)
 		if err != nil {
 			// TODO: Trigger network garbage collection
 			// TODO: Check if all IPs in rethinkdb are in the IPAM and vice versa, cleanup if this is not the case
 			// TODO: Check if there are network prefixes in the IPAM that are not in any of our networks
-			logger.Errorw("an error during releasing machine networks occurred, scheduled network garbage collection", "error", err)
+			logger.Errorf("an error during releasing machine networks occurred, scheduled network garbage collection", "error", err)
```
Comment (Gerrit91) [inline cmd/metal-api/internal/service/machine-service.go]: This must be an errorw as before, otherwise you get bad formatting error in the output.

File: cmd/metal-api/internal/service/machine-service.go [hunk 61]
```diff
@@ -1555,6 +1578,129 @@ func freeMachine(ds *datastore.RethinkStore, publisher bus.Publisher, ipamer ipa
...
+				return
+			}
+
+			err = publishMachineCmd(logger, m, r, metal.MachineReinstall)
+			if err != nil {
+				logger.Errorw("unable to publish machine command", "command", metal.MachineReinstall, "machineID", m.ID, "error", err)
+			}
+		}
+	} else {
+		resp = makeMachineResponse(m, r.ds, logger)
```
Comment (Gerrit91) [inline cmd/metal-api/internal/service/machine-service.go]: So... you can call this endpoint and when there is no allocation just nothing happens and the result is "ok"? Wouldn't it be better to return an error if there is no allocation?
